edge/pkg/devicetwin/dmiclient/client.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2022 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dmiclient
+
+import (
+	"net"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"k8s.io/klog/v2"
+
+	deviceconst "github.com/kubeedge/kubeedge/cloud/pkg/devicecontroller/constants"
+	dmiapi "github.com/kubeedge/kubeedge/edge/pkg/apis/dmi/v1"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
+	"github.com/kubeedge/kubeedge/pkg/apis/devices/v1alpha2"
+)
+
+func GenerateDMIClient(sockPath string) (dmiapi.DeviceMapperServiceClient, context.Context, *grpc.ClientConn, context.CancelFunc, error) {
+	klog.Infof("@@@ generate DMI client")
+	dialer := func(addr string, t time.Duration) (net.Conn, error) {
+		return net.Dial(deviceconst.UnixNetworkType, addr)
+	}
+
+	klog.Infof("@@@ Dial DMI client")
+	conn, err := grpc.Dial(sockPath, grpc.WithInsecure(), grpc.WithDialer(dialer))
+	if err != nil {
+		klog.Errorf("did not connect: %v\n", err)
+		return nil, nil, nil, nil, err
+	}
+
+	klog.Infof("@@@ new DMI client")
+	c := dmiapi.NewMapperClient(conn)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+
+	return c, ctx, conn, cancel, nil
```

Comment:
can wrap a struct and a close function to make it more readable, like
```
type client struct {
	client dmiapi.DeviceMapperServiceClient
	context context.Context
	conn *grpc.ClientConn
	cancel context.CancelFunc
}
func (c *client) Close() {
	c.conn.Close()
	c.cancel()
}
```

Reply:
and can also cal CreateDeviceModel of interface dmiapi.DeviceMapperServiceClient
```
a := &client{}
defer a.Close()
a.CreateDeviceModel(nil, nil)
```

Reply:
good idea. make it.

> can wrap a struct and a close function to make it more readable, like
> 
> ```
> type client struct {
> 	dmiapi.DeviceMapperServiceClient
> 	context context.Context
> 	conn *grpc.ClientConn
> 	cancel context.CancelFunc
> }
> func (c *client) Close() {
> 	c.conn.Close()
> 	c.cancel()
> }
> ```

Reply:
for this suggestion, i have make the dmi client as a map of clients and create the client only when a mapper register. thus we can just get the client from map and do not need to create a tmp client before `CreateDeviceModel`.

> and can also call function `CreateDeviceModel` of interface `dmiapi.DeviceMapperServiceClient` directively
> 
> ```
> a := &client{}
> defer a.Close()
> a.CreateDeviceModel(nil, nil)
> ```